arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt

```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
    *  ```
    */
   fun <B> bracketCase(use: (A) -> MonoKOf<B>, release: (A, ExitCase<Throwable>) -> MonoKOf<Unit>): MonoK<B> =
-    MonoK(Mono.create<B> { sink ->
-      val isCancelled = AtomicBooleanW(false)
-      sink.onCancel { isCancelled.value = true }
-      val a: A? = mono.block()
-      if (a != null) {
-        if (isCancelled.value) release(a, ExitCase.Cancelled).fix().mono.subscribe({}, sink::error)
-        else try {
-          sink.onDispose(use(a).fix()
-            .flatMap { b -> release(a, ExitCase.Completed).fix().map { b } }
-            .handleErrorWith { e -> release(a, ExitCase.Error(e)).fix().flatMap { MonoK.raiseError<B>(e) } }
-            .mono
-            .doOnCancel { release(a, ExitCase.Cancelled).fix().mono.subscribe({}, sink::error) }
-            .subscribe(sink::success, sink::error)
+    Mono.create<B> { emitter ->
+      val cancelable = ForwardDisposable()
+      emitter.onCancel(cancelable.cancel())
+
+      value().subscribe(
+        { a: A ->
+          cancelable.complete(
+            Mono.defer { use(a).value() }
+              .doOnCancel { Mono.defer { release(a, ExitCase.Cancelled).value() }.subscribe() }
+              .subscribe({ b ->
+                Mono.defer { release(a, ExitCase.Completed).value() }
+                  .subscribe({ emitter.success(b) }, { e -> emitter.error(e) })
+              }, { e ->
+                Mono.defer { release(a, ExitCase.Error(e)).value() }
+                  .subscribe({ emitter.error(e) }, { e2 -> emitter.error(Platform.composeErrors(e, e2)) })
+              }, {
+                Mono.defer { release(a, ExitCase.Completed).value() }
+                  .subscribe({ emitter.success() }, { e -> emitter.error(e) })
+              })
           )
-        } catch (e: Throwable) {
-          if (NonFatal(e)) {
-            release(a, ExitCase.Error(e)).fix().mono.subscribe({
-              sink.error(e)
-            }, { e2 ->
-              sink.error(Platform.composeErrors(e, e2))
-            })
-          } else {
-            throw e
-          }
-        }
-      } else sink.success(null)
-    })
+        },
+        emitter::error,
+        { }
```

Comment:
completion? do they need to be forwarded?

Reply:
No, the inner subcribed `Mono` calls the terminal operators.